File: demo-server/src/ws/manager.rs
use std::{collections::HashMap, sync::Arc};

use tokio::sync::Mutex;

use super::session::WsSession;

#[derive(Clone, Default)]
pub struct WsSessionManager {
    sessions: Arc<Mutex<HashMap<uuid::Uuid, WsSession>>>,
}

impl WsSessionManager {
    pub async fn get_session(&self, id: &uuid::Uuid) -> Option<WsSession> {
        let sessions = self.sessions.lock().await;
        sessions.get(id).cloned()
    }

    pub async fn add_session(&self, session: WsSession) {
        let mut sessions = self.sessions.lock().await;
        sessions.insert(session.id, session);
    }

    pub async fn remove_session(&self, session: WsSession) {
        let mut sessions = self.sessions.lock().await;
        sessions.remove(&session.id);
    }
}

File: fastapi-server/app/ws/manager.py
import uuid
from typing import Dict, Optional

from app.ws.session import WsSession


class WsSessionManager:
    _singleton_instance = None

    def __init__(self):
        self.sessions: Dict[uuid.UUID : WsSession] = {}

    @classmethod
    def instance(cls):
        if not cls._singleton_instance:
            cls._singleton_instance = WsSessionManager()
        return cls._singleton_instance

    async def get_session(self, id: uuid.UUID) -> Optional[WsSession]:
        return self.sessions.get(id, None)

    async def add_session(self, session: WsSession):
        self.sessions[session.id] = session

    async def remove_session(self, id: uuid.UUID):
        self.sessions.pop(id, None)
